Local implementation

freddie-freeloader opened the pull request:
WIP

freddie-freeloader commented on compiler/src/dotty/tools/dotc/Compiler.scala:
```diff
@@ -33,9 +33,10 @@ class Compiler {
     picklerPhases :::
     transformPhases :::
+    List(List(new DenoteSafeParams)) :::
+    List(List(new TestCheck)) :::
```
These classes are artifacts of the annotation based version.

freddie-freeloader commented on compiler/src/dotty/tools/dotc/parsing/Parsers.scala:
```diff
@@ -1973,6 +1987,8 @@ object Parsers {
           }
           for (imod <- imods.mods) mods = addMod(mods, imod)
+          if (imods.hasLocalQualifier)
+            mods = mods.withLocalQualifier(imods.localQualifier)
```
Modifiers were copied quite selectively, so we have to copy `localQualifier` explicitly. :disappointed:

freddie-freeloader commented on compiler/src/dotty/tools/dotc/typer/Typer.scala:
```diff
@@ -755,26 +755,29 @@ class Typer extends Namer
 
   def typedFunctionType(tree: untpd.Function, pt: Type)(implicit ctx: Context): Tree = {
```
Sadly this constructs a tree out of our beautiful `FunctionWithMods`. The tree is something like `AppliedType(TypeConstructor=LocalImplicitFunction1,Int,Int)`).
Therefore, it is not obvious where to store `localQualifier`.

freddie-freeloader commented on compiler/src/dotty/tools/dotc/typer/Typer.scala:
```diff
@@ -1947,7 +1950,7 @@ class Typer extends Namer
   }
 
   protected def makeImplicitFunction(tree: untpd.Tree, pt: Type)(implicit ctx: Context): Tree = {
```
Sadly here only the `AppliedType` (as `pt`) arrives.
It gets deconstructed by `defn.Function` which simply parses the class of the type constructor. :disappointed:

freddie-freeloader replied:
Should `localQualifier` be one of the arguments of `LocalFunctionN`?

This could probably be done in `newFunctionNTrait` but looks far from trivial.

b-studios commented on compiler/src/dotty/tools/dotc/core/StdNames.scala:
```diff
@@ -35,7 +35,9 @@ object StdNames {
     final val ImplicitFunction           = "ImplicitFunction"
+    final val LocalFunction              = "LocalFunction"
     final val ErasedImplicitFunction     = "ErasedImplicitFunction"
+    final val LocalImplicitFunction      = "LocalImplicitFunction"
```
What about
```suggestion
    final val LocalImplicitFunction      = "LocalImplicitFunction"
    final val LocalErasedImplicitFunction      = "LocalErasedImplicitFunction"
```
?

b-studios replied:
Or
```suggestions
LocalImplicitFunction[A...]
```

with 

```scala
trait LocalImplicitFunction[-L, -A, +B] {
  def apply(implicit local[L] arg: A): B
}
```

b-studios commented on compiler/src/dotty/tools/dotc/core/Types.scala:
```diff
@@ -2984,8 +2988,12 @@ object Types {
+      companion.eq(ErasedImplicitMethodType) ||
+      (companion match { case LocalMethodType(_,isImplicit) => isImplicit; case _ => false })
     final override def isErasedMethod: Boolean = companion.eq(ErasedMethodType) || companion.eq(ErasedImplicitMethodType)
+    final override def isLocalMethod: Boolean = companion.isInstanceOf[LocalMethodType]
```
Why is this different from `isErasedMethod`?

b-studios replied:
Ah, I see.. it is a case class, not a case object...

b-studios commented on compiler/src/dotty/tools/dotc/parsing/Parsers.scala:
```diff
@@ -1794,14 +1795,26 @@ object Parsers {
+        //if ((mods is LocalMod) || mods.hasLocalQualifier)
+        //  syntaxError(DuplicatePrivateProtectedQualifier())
+        inBrackets {
+          mods.withLocalQualifier(ident().toTypeName)
```
You probably want to parse a whole type here, not just a name.

For instance:

```scala
def foo(local[A with B] n: Int) ...
```

b-studios commented on compiler/src/dotty/tools/dotc/core/SymDenotations.scala:
```diff
@@ -723,7 +734,7 @@ object SymDenotations {
         || boundary.isRoot
         || (accessWithin(boundary) || accessWithinLinked(boundary)) &&
-             (  !(this is Local)
+             (  !(this is LocalAccess)
```
Ah, I see this is an access modifier flag...

b-studios commented on compiler/src/dotty/tools/dotc/core/Symbols.scala:
```diff
@@ -58,9 +58,10 @@ trait Symbols { this: Context =>
       coord: Coord = NoCoord): Symbol { type ThisName = N } = {
     val sym = newNakedSymbol[N](coord)
-    val denot = SymDenotation(sym, owner, name, flags, info, privateWithin)
+    val denot = SymDenotation(sym, owner, name, flags, info, privateWithin, localQualifier)
```
What about dropping `localQualifier` for now to make it simpler?

b-studios commented:
I am curious, did you manage to drop the `localQualifier`? :)

freddie-freeloader commented:
Yes, I did that. I'll create a new branch and pull request soon. :slightly_smiling_face:  I fear a rebase would make the comments in this PR disappear.

b-studios commented:
They won't disappear, but show up as "There are comments on older versions" or something like this

freddie-freeloader committed:
First steps of `dotty-escape` implementation

- Add support in parser for `local` modifier
- Add needed denotation
- Add new method companions `LocalMethodType` and
  `LocalImplicitMethodType`
- Add first tests
- Add first rules of the checker (in `EscapeCheck`)

freddie-freeloader commented:
I squashed the commits so the removal of `localQualifier` does not pollute the commits too much. I hope that is okay. I've got the individual commits saved  on another branch just in case + for the sake of nostalgia. :slightly_smiling_face:

freddie-freeloader commented:
The only "new" stuff:
- Some initial code for the new checker/traverser in [`EscapeCheck.scala`](https://github.com/freddie-freeloader/dotty/blob/4bfe302f9bffa87344fe548af0b944ab1b2ecb76/compiler/src/dotty/tools/dotc/escapecheck/EscapeCheck.scala)
- Some basics test cases in [`tests/neg/local-modifier-1.scala`](https://github.com/freddie-freeloader/dotty/blob/4bfe302f9bffa87344fe548af0b944ab1b2ecb76/tests/neg/local-modifier-1.scala)

b-studios commented:
Great that you could extract the local-check to a separate traversal!

b-studios commented on compiler/src/dotty/tools/dotc/escapecheck/EscapeCheck.scala:
```diff
@@ -0,0 +1,97 @@
+
+
+      case Block(stats, expr) =>
+        stats.foreach(traverseWithLocalMode(_, secondClass))
```
Why do you traverse with `secondClass` and not with the current classiness?

freddie-freeloader replied:
```scala
def foo(local x: Int) = {
  x
  42
}
```

b-studios replied:
Yes, you're right.

b-studios commented on compiler/src/dotty/tools/dotc/escapecheck/EscapeCheck.scala:
```diff
@@ -0,0 +1,97 @@
+        traverseWithLocalMode(tree.rhs, getClassiness(tree))
+
+      case tree : DefDef =>
+        traverse(tree.rhs)(ctx.fresh.setLocalMode(firstClass).addBoundary(tree.symbol))
```
Is there something like local functions so you could also `traverseWithLocalMode` ?

b-studios commented on compiler/src/dotty/tools/dotc/escapecheck/EscapeCheck.scala:
```diff
@@ -0,0 +1,97 @@
+        }
+
+
+      case Block(stats @ (fn: DefDef) :: Nil, Closure(_, fnRef, tpt)) if fnRef.symbol == fn.symbol =>
```
Does this pattern also cover the case of local function definitions like
```scala
{
  local def foo() : ...
  local def bar() : ...
  f(foo, bar)
}
```
?

b-studios replied:
Is this true?
```scala
val f = x => BODY
```
desugars to
```scala
val f = {
  def anon(x) { BODY}
  anon
}
```

with local:

```scala
local val f = {
  def anon(x) { BODY}
  anon
}
```

So, if we are in local mode when typechecking the closure, just annotate `anon`:

```scala
local val f = {
  local def anon(x) { BODY}
  anon
}
```

and then typecheck the block as before (in second mode)

freddie-freeloader replied:
From `Desugar.scala`:
```scala
    /** Make closure corresponding to function.
   *      params => body
   *  ==>
   *      def $anonfun(params) = body
   *      Closure($anonfun)
   */
  def makeClosure(params: List[ValDef], body: Tree, tpt: Tree = TypeTree(), isImplicit: Boolean)(implicit ctx: Context): Block =
    Block(
      DefDef(nme.ANON_FUN, Nil, params :: Nil, tpt, body).withMods(synthetic | Artifact),
      Closure(Nil, Ident(nme.ANON_FUN), if (isImplicit) ImplicitEmptyTree else EmptyTree))
```

b-studios commented on compiler/src/dotty/tools/dotc/escapecheck/EscapeCheck.scala:
```diff
@@ -0,0 +1,97 @@
+
+      case Block(stats, expr) =>
+        stats.foreach(traverseWithLocalMode(_, secondClass))
+        traverseWithLocalMode(expr, firstClass)
```
~Oops, this should be `currentMode`, not `firstClass`~

freddie-freeloader commented:
Fun fact:

This is invalid code:
```scala
def f(@local x: Int) = 2 + x
```

And this is valid code:
```scala
def f(@local x: Int) = x + 2
```

:laughing:

freddie-freeloader commented:
I wonder whether we should also traverse the type trees. This might be relevant because of path dependent types:

```scala
class HazType() { type T }
local val iHazType = new HazType() { type T = Int }
def f(x: iHazType.T): iHazType.T = x
type LookAtMeFromAnotherScope = iHazType.T
```

I am not quite sure what to do here.

freddie-freeloader committed:
Add first draft of safe modifier

freddie-freeloader committed:
Fix some bugs for `local` in FrontEnd

freddie-freeloader committed:
Extend `EscapeCheck`

b-studios commented:
Regarding `2 + x` vs `x + 2`. This problem will always show up with value-like types that easily can escape by copying: e.g. `x + 0` is first class.

The question what does it mean to call a method on a second class value should maybe better be asked in a context where we don't necessarily need to be afraid of copying.

freddie-freeloader committed:
Add basic implicit local function expansion

freddie-freeloader commented on tests/neg/local-implicit-function.scala:
```diff
@@ -0,0 +1,45 @@
+      throwing(new Exception) // error
+    }
+  }
+}
```
This example is based on one from *the gentrification paper*.
With the help of implicit (local) function types this is now even niftier.

Old:
```scala
trying { throwing =>
  runUnsafe {
    () => throwing(new Exception) // error
   }
}
```

New:
```scala
trying {
  runUnsafe {
    throwing(new Exception) // error
   }
}
```
:tada:

b-studios replied:
That's what I am talking about ;)

b-studios replied:
Currently bootstrapping the Dotty compiler fails, since `local` is used as a name, for instance in `Desugar.scala:90:16`

```scala
var local = defctx.denotNamed(tp.name ++ suffix).suchThat(_.isParamOrAccessor).symbol
//  ^^^^^
```

freddie-freeloader replied:
Good point!

Should we try to demote `local` and `safe` from keywords to identifiers for better compatibility?

b-studios replied:
Not sure what's the best strategy here, but I would be fine with keeping them as keywords and renaming usages to `l` or something else.

b-studios replied:
(As long as the renaming happens in one commit ;) )

b-studios replied:
Another side-note: It looks like `local` makes currying a bit akward:

```scala
f: implicit local A => implicit local B => C
//   =?=
g: implicit local (A, B) => C
```
in function `f` we will close over the local of type `A`, which will be rejected. And rightfully so (in the current type system)! applying `f` to only `a` gives us a *first class* function closing over `a`.

b-studios commented on tests/neg/local-implicit-function.scala:
```diff
@@ -0,0 +1,45 @@
+    def incr(i2: Int) = i1 + i2
+  }
+
+  def incr(i: Int): Int using Incrementor = implicitly[Incrementor].incr(i)
```
Why does `implicitly` work at all? isn't it defined as
```scala
def implicitly[A](implicit a: A): A = a
```
?
That would require `a` to be first class, doesn't it?

freddie-freeloader replied:
Yes, `implicitly` breaks `local` somehow.

I investigated this some time ago and my suspicion is that `@forceInline` in the definition of `implicitly` has something to do with that.

```scala
@forceInline final def implicitly[T](implicit ev: T): T = ev
```

So I tested it now:

```scala
object Test {

  implicit local val foo: Int = 42
  @forceInline final def explicitly[T](         ev: T): T = ev
  @forceInline final def implicitly[T](implicit ev: T): T = ev
  local val bar  = explicitly(foo)
  local val cafe = implicitly[Int]

}
```

This works. So the code seems to get inlined before our phase.

b-studios replied:
Maybe that's a feature and not a bug :)

b-studios replied:
(as long as `val bar = implicitly[Int]` doesn't type check)

b-studios replied:
Brilliant!

```scala
  local val x = 42
  var y: Int = _
  @forceInline final def tester(ev: Int): Unit = y = ev
  tester(x)
  //  |  @forceInline final def tester(ev: Int): Unit = y = ev
  //  |                                                     ^^
  //  |                            Found local value x in first class position.
  //  | This location is in code that was inlined at examples/test.scala:49
```

b-studios commented:
I played around with your implementation of `local`.
Here is an example of implementing control-effects. I assume that
we have a runtime that supports delimited control (without being monadic), but hide it behind the library interface. Such a runtime currently doesn't exist (for multiple resumptions) hence the example is only for type checking.

```scala
object effekt {
  // LIBRARY INTERFACE
  class Prompt private ()

  trait Handler[R] {
    type Effects

    // we need to prevent something like: new Maybe{}.raise("foo")
    // or div(4, 0)(new Maybe) since `Maybe` uses delimited continuations.
    erased val p: Prompt

    // the continuation is not second class. However,
    // even if it escapes, it can only be used in the scope where
    // this.Effects is available...
    type Continuation[-A, +B] = A => implicit local Effects => B
    def use[A](body: Continuation[A, R] => implicit local Effects => R): A = ???
  }
  // passing the erased prompt here imposes some boilerplate on handler creation ...
  def handle[R, H <: Handler[R]](h: erased Prompt => H)(local prog: implicit local H => R): R = ???

  // USER CODE
  trait Exc { def raise(msg: String): Nothing }
  trait Amb { def flip(): Boolean }

  def div(x: Int, y: Int)(implicit local exc: Exc): Int =
    if (y == 0) exc raise "division by zero"
    else x / y

  def drunkFlip(implicit local amb: Amb, exc: Exc): String =
    if (!amb.flip()) {
      exc raise "too drunk"
    } else if (amb.flip()) {
      "Heads"
    } else {
      "Tails"
    }

  trait Collect[R] extends Amb with Handler[List[R]] {
    def flip() = use { resume => resume(true) ++ resume(false) }
  }
  def Collect[R](local prog: implicit local Amb => R): List[R] =
    handle(erased prompt => new Collect[R] { erased val p = prompt }) {
      List(prog)
    }

  trait Maybe[R] extends Exc with Handler[Option[R]] {
    def raise(msg: String) = use { resume => None }
  }
  def Maybe[R](local prog: implicit local Exc => R): Option[R] =
    handle(erased prompt => new Maybe[R] { erased val p = prompt }) {
      Some(prog)
    }

  val res1: Option[List[String]] = Maybe {
    Collect {
      drunkFlip
    }
  }

  val res2: List[Option[String]] = Collect {
    Maybe {
      drunkFlip
    }
  }
}
```

b-studios commented:
Not that you need to implement that, but just to not forget it: Disregarding many details (subtyping and so forth), we can think of an object (with multiple methods) as a straightforward generalization of a function value (In fact, both are _"just"_ codata). As such, a local object should be able to close over other local values in each of its methods. For instance

```scala
  local val x: Int = ???
  def f1(local g: () => Unit) = ???
  // works
  f1 { () => println(x + 0) }

  trait Foo { def foo(): Unit; }
  def f2(local g: Foo) = ???

  // still works because of SAMs
  f2 { () => println(x + 0) } 
  
  // !!! doesn't work
  f2 { new Foo { def foo() = println(x + 0) } }
```
This should also hold for `erased`, btw.

freddie-freeloader commented:
In contrary to the `implicitly` bug-feature, this seems a bit nasty:

```scala
scala> def trusted(local nl: NukeLauncher): () => Unit  = {
     |   def mb(local nl: NukeLauncher): Int = { nl.launch(); 42 }
     |   lazy val danger = mb(nl)
     |   () => danger
     | }
def trusted(nl: NukeLauncher): () => Unit

scala> val dd = trusted(nl)
val dd: () => Unit = Lambda$1675/218857805@6ce15112

scala> dd()                                                                                                                                                                   
Boooooooooooooooooooooooooooom
```

b-studios commented:
Yes, that's bad. Lazy is like a nullary (first-class), auto-applied function and should not close over locals.

Exceeeept:
```scala
local lazy val foo = println(myLocal + 0)
```

b-studios commented:
Sorry for bringing up so many things at once. Here is another surprising example:

```scala
// type checks
class Foo1(local val y: Int) {
  local val x = y
  def foo = println(x + 0)
}
// vs.
// doesn't typecheck
class Foo2(local val x: Int) {
  def foo = println(x + 0)
}
```

even worse, the following also typechecks:
```scala
class Foo1(local val y: Int) {
  local val x = y
  def foo = x
}
```

freddie-freeloader committed:
Add multiple stuff

freddie-freeloader committed:
Check function definitions for local parameters

- Local parameters are only allowed to appear in the last parameter list
  - Add a check for that

freddie-freeloader committed:
Make README project specific

freddie-freeloader commented on compiler/src/dotty/tools/dotc/escapecheck/EscapeCheck.scala:
```diff
@@ -72,10 +72,28 @@ class EscapeCheck extends Phase {
+          if (isSecondClass(vparam)
+              && (i + 1) < tree.vparamss.length) {
+            ctx.error(
+              em"Local parameters are only allowed to appear in the last parameter list."
```
This is actually over restrictive. Since we have `erased` in our language, we could allow `erased local` parameters for any position. :blush:
